feat(support-mysql-8): Added support for mysql version 8.0 and above

sairaj18 opened the pull request:
The query cache variables are removed from MySQL 8.0 - https://dev.mysql.com/doc/refman/5.7/en/query-cache-status-and-maintenance.html - Due to this the following metrics will not be support for mysql version 8.0 and above

1. db.qCacheFreeBlocks
2. db.qCacheFreeMemoryBytes
3. db.qCacheHitRatio
4. db.qCacheHitsPerSecond
5. db.qCacheInserts
6. db.qCacheLowmemPrunesPerSecond
7. db.qCacheNotCachedPerSecond
8. db.qCacheQueriesInCachePerSecond
9. db.qCacheTotalBlocks
10. db.qCacheUtilization

From Mysql 8.4 SHOW SLAVE STATUS is removed and SHOW REPLICA STATUS should be used instead
Ref - https://dev.mysql.com/doc/relnotes/mysql/8.4/en/news-8-4-0.html#:~:text=SQL%20statements%20removed

Code changes -
- Updated the getReplicaQuery func to return query based on mysql version
- Updated the metric definitions to handle version based metrics
- Updated, Added unit tests for the same
- Added integration tests for mysql versions 8.0.40 and 9.1.0

Testing details [here](https://new-relic.atlassian.net/browse/NR-133316?focusedCommentId=1090401)

rajrohanyadav commented on src/metrics-parse.go:
```diff
@@ -1,20 +1,36 @@
-	replicaQuery   = "SHOW SLAVE STATUS"
+	inventoryQuery  = "SHOW GLOBAL VARIABLES"
+	metricsQuery    = "SHOW /*!50002 GLOBAL */ STATUS"
+	replicaQuery560 = "SHOW SLAVE STATUS"
```
Is this only for version `5.6.0`? I'd assume this will be same for `5.7.x` also. If so, can we update the name to be a bit more descriptive. Right now it seems that this will only work for `5.6.0`.

sairaj18 replied:
`replicaQueryUnder800` ?

sairaj18 replied:
it should be `replicaQueryBelowVersion8`

sairaj18 replied:
Resolved it in this [commit](https://github.com/newrelic/nri-mysql/pull/170/commits/fe29f92e5ad2d923fd031d278d2ecb482b0e69be)

rajrohanyadav commented on src/metrics-parse.go:
```diff
@@ -1,20 +1,36 @@
+	inventoryQuery  = "SHOW GLOBAL VARIABLES"
+	metricsQuery    = "SHOW /*!50002 GLOBAL */ STATUS"
+	replicaQuery560 = "SHOW SLAVE STATUS"
+	replicaQuery800 = "SHOW REPLICA STATUS"
```
> Is this only for version 5.6.0? I'd assume this will be same for 5.7.x also. If so, can we update the name to be a bit more descriptive. Right now it seems that this will only work for 5.6.0.

Similar point here.

sairaj18 replied:
this should be `replicaQueryForVersion8AndAbove`

sairaj18 replied:
Resolved it in this [commit](https://github.com/newrelic/nri-mysql/pull/170/commits/fe29f92e5ad2d923fd031d278d2ecb482b0e69be)

rajrohanyadav commented on src/default_metrics.go:
```diff
@@ -0,0 +1,127 @@
+	"cluster.nodeType":                            {"node_type", metric.ATTRIBUTE},
+}
+
+var defaultMetrics560 = map[string][]interface{}{
```
> Is this only for version 5.6.0? I'd assume this will be same for 5.7.x also. If so, can we update the name to be a bit more descriptive. Right now it seems that this will only work for 5.6.0.

Same here.

sairaj18 replied:
Resolved it in this [commit](https://github.com/newrelic/nri-mysql/pull/170/commits/fe29f92e5ad2d923fd031d278d2ecb482b0e69be)

rajrohanyadav commented on src/default_metrics.go:
```diff
@@ -0,0 +1,127 @@
+	"cluster.slaveRunning": {slaveRunningAsNumber, metric.GAUGE},
+}
+
+func slaveRunningAsNumber(metrics map[string]interface{}, version *semver.Version) (int, bool) {
```
I think this requires unit tests.

sairaj18 replied:
This is already existing func just moved to diff file

rajrohanyadav commented on src/default_metrics.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+func qCacheUtilization(metrics map[string]interface{}) (float64, bool) {
+	// TODO compute the value within the interval
```
This TODO is not clear to me.

sairaj18 replied:
this is already existing code that just moved to different file

rajrohanyadav replied:
We need to understand why it is there. If we think we don't need it, it can be removed.

rajrohanyadav commented on src/default_metrics.go:
```diff
@@ -0,0 +1,127 @@
+
+	if queries == 0 {
+		return 0, true
+	} else if ok1 && ok2 {
```
Removing `else` here improves code readability IMO.

sairaj18 replied:
Resolved it in this [commit](https://github.com/newrelic/nri-mysql/pull/170/commits/fe29f92e5ad2d923fd031d278d2ecb482b0e69be)

rajrohanyadav commented on src/default_metrics.go:
```diff
@@ -0,0 +1,127 @@
+var defaultMetricsVersionDefinitions = []VersionDefinition{
+	{
+		minVersion:        semver.MustParse("8.0.0"),
+		metricsDefinition: mergeMaps(defaultMetricsBase, defaultMetrics800),
```
Can we use struct embedding rather than merging structures?

sairaj18 replied:
proceeding with using mergeMaps as we cannot use embedding with maps. 

Resolved it in this [commit](https://github.com/newrelic/nri-mysql/pull/170/commits/fe29f92e5ad2d923fd031d278d2ecb482b0e69be)

rajrohanyadav commented on src/default_metrics.go:
```diff
@@ -0,0 +1,127 @@
+	},
+	{
+		minVersion:        semver.MustParse("5.6.0"),
+		metricsDefinition: mergeMaps(defaultMetricsBase, defaultMetrics560),
```
Can we use struct embedding rather than merging structures?

sairaj18 replied:
proceeding with using mergeMaps as we cannot use embedding with maps. 

Resolved it in this [commit](https://github.com/newrelic/nri-mysql/pull/170/commits/fe29f92e5ad2d923fd031d278d2ecb482b0e69be)

rajrohanyadav commented on src/extended_metrics.go:
```diff
@@ -0,0 +1,77 @@
+	},
+	{
+		minVersion:        semver.MustParse("5.6.0"),
+		metricsDefinition: mergeMaps(extendedMetricsBase, extendedMetrics560),
```
Can we use struct embedding rather than merging structures?

sairaj18 replied:
proceeding with using mergeMaps as we cannot use embedding with maps. 

Resolved it in this [commit](https://github.com/newrelic/nri-mysql/pull/170/commits/fe29f92e5ad2d923fd031d278d2ecb482b0e69be)

rajrohanyadav commented on src/extended_metrics.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+func threadCacheMissRate(metrics map[string]interface{}) (float64, bool) {
+	// TODO compute the value within the interval
```
I didn't understand the TODO statement here.

sairaj18 replied:
this is already existing comment that just moved to different file

rajrohanyadav replied:
We need to understand why it is there. If we think we don't need it, it can be removed.

rajrohanyadav commented on src/metrics-parse.go:
```diff
@@ -126,3 +151,53 @@ func populatePartialMetrics(ms *metric.Set, metrics map[string]interface{}, metr
+}
+
+// extractSemver uses a regular expression to extract a semver-compliant version string up to major.minor.patch
+func extractSemver(version string) (string, error) {
+	reg := regexp.MustCompile(`^(?P<major>\d+)(?:\.(?P<minor>\d+))?(?:\.(?P<patch>\d+))?`)
+	matches := reg.FindStringSubmatch(version)
+
+	if len(matches) == 0 || matches[1] == "" {
+		return "", errSemanticVersionNotFound
+	}
+
+	major := matches[1]
+	minor := "0"
+	patch := "0"
+
+	if len(matches) > 2 && matches[2] != "" {
+		minor = matches[2]
+	}
+	if len(matches) > 3 && matches[3] != "" {
+		patch = matches[3]
+	}
+
+	return fmt.Sprintf("%s.%s.%s", major, minor, patch), nil
+}
```
I don't think we need this much complexity for getting the version. [MySQL Doc](https://dev.mysql.com/doc/refman/8.4/en/information-functions.html#function_version) specifies that the version might have a suffix but not a prefix. Therefore, you'll always get version before the first `-`.
Moreover, the documentation states: `This function is unsafe for statement-based replication. A warning is logged if you use this function when [binlog_format](https://dev.mysql.com/doc/refman/8.4/en/replication-options-binary-log.html#sysvar_binlog_format) is set to STATEMENT.` Though the `binlog_format` is deprecated, do you think it might cause any issues for older customer environments?

sairaj18 replied:
will discuss this with kai

sairaj18 replied:
I ran integration tests in github actions for version 5.7.35 with binlog_format=STATEMENT and it [passed](https://github.com/newrelic/nri-mysql/actions/runs/12559212682/job/35014599864?pr=170). I didn’t make any change to handle if we are not able to parse DB version output. So, We are getting the version as excepted without any warning with was mentioned in the Mysql doc.

rajrohanyadav replied:
I think we should just note this in the documentation. Other than that it seems fine.

rajrohanyadav commented on src/slave_metrics.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+// mergeMaps merges two maps of type map[string][]interface{}
+func mergeMaps(map1, map2 map[string][]interface{}) map[string][]interface{} {
+	merged := make(map[string][]interface{})
+	for k, v := range map1 {
+		merged[k] = v
+	}
+	for k, v := range map2 {
+		merged[k] = v
+	}
+	return merged
+}
```
We should try to not do this :(
See my other comment about using embedding rather than doing this. Even if there is no option and we need to use this, you can always keep one map and copy over the shorter map, rather than creating a new one and copying both ones.

sairaj18 replied:
Updated the func to overwrite map2 into map1 instead of creating new map

Resolved it in this [commit](https://github.com/newrelic/nri-mysql/pull/170/commits/fe29f92e5ad2d923fd031d278d2ecb482b0e69be)

rajrohanyadav reviewed (changes requested):
Thanks for taking care of this.

Left some comments. Some more points:
1. Also, a lot of new functions are added without unit tests. I think it makes sense to add tests for the ones that have logical statements.
2. Do you think it improves the code if we have an optional argument in config with db version? We won't have to compute version on each execution. If not specified, we anyways do it.
3. IMO, the additional dependency is not required. But we can discuss.
4. Can `version` throughout the code be replaced with `dbVersion` or something similar. It could be confused with integration version.

sairaj18 commented on Makefile:
```diff
@@ -23,12 +23,12 @@ compile: bin/$(BINARY_NAME)
 test:
 	@echo "=== $(INTEGRATION) === [ test ]: running unit tests..."
-	@go test -race ./... -count=1
+	@go test -v -race ./... -count=1
```
I was not able to know if the tests have ran or not. So, added the verbose to know if test ran and passed/failed.

sairaj18 replied:
Resolved it in this [commit](https://github.com/newrelic/nri-mysql/pull/170/commits/fe29f92e5ad2d923fd031d278d2ecb482b0e69be)Resolved this in commit

sairaj18 commented on Makefile:
```diff
@@ -23,12 +23,12 @@ compile: bin/$(BINARY_NAME)
 	@echo "=== $(INTEGRATION) === [ test ]: running integration tests..."
 	@docker compose -f tests/integration/docker-compose.yml up -d --build
-	@go test -tags=integration ./tests/integration/. || (ret=$$?; docker compose -f tests/integration/docker-compose.yml down && exit $$ret)
+	@go test -v -tags=integration ./tests/integration/. || (ret=$$?; docker compose -f tests/integration/docker-compose.yml down && exit $$ret)
```
Same here as well I was not able to know if the tests have ran or not. So, added the verbose to know if test ran and passed/failed.

sairaj18 replied:
Resolved it in this [commit](https://github.com/newrelic/nri-mysql/pull/170/commits/fe29f92e5ad2d923fd031d278d2ecb482b0e69be)

sairaj18 commented on go.mod:
```diff
@@ -4,6 +4,7 @@ go 1.23.4
 
 require (
 	github.com/bitly/go-simplejson v0.5.1
+	github.com/blang/semver/v4 v4.0.0
```
The same dependency was used by rahul in nri-promethus [ref](https://github.com/newrelic/nri-postgresql/blob/0de146486c5e111e32f01c7950493bff3f7a9c04/go.mod#L6C2-L6C35) to compare diff DB versions So, used it here as well. We could use simple regex as well but as per ref I followed the same.

sairaj18 replied:
Resolved it in this [commit](https://github.com/newrelic/nri-mysql/pull/170/commits/fe29f92e5ad2d923fd031d278d2ecb482b0e69be)

sairaj18 commented on src/default_metrics.go:
```diff
@@ -0,0 +1,127 @@
+package main
```
only `metrics-parse.go` is using `-`. In other repos we are using `_`. So, Do you suggest renaming `metrics-parse.go` -> `metrics_parse.go` or renaming as the files that i have added with `_` -> `-` ?

sairaj18 replied:
for now we are going ahead with `_` instead of `-`

sairaj18 commented on src/metrics-parse.go:
```diff
@@ -1,20 +1,36 @@
+	versionQuery    = "SELECT VERSION() as version;"
 )
 
+var errVersionNotFound = errors.New("version not found in versionQueryResult")
+var errSemanticVersionNotFound = errors.New("no semantic version found")
```
yep will make changes for this.

sairaj18 replied:
look like we cannot assign non-constant type error to a constant

sairaj18 replied:
Error: do not define dynamic errors, use wrapped static errors instead: "errors.New(versionNotFound)" (err113)
  Error: do not define dynamic errors, use wrapped static errors instead: "errors.New(semanticVersionNotFound)" (err113)
  
Getting the above error so creating var in the module level

rajrohanyadav replied:
FWIW, I don't agree with this recommendation. We don't need allocation of a new error which could never be used. This pattern maybe makes sense in a microservice, but not here. I'd have just ignored the linting error :)

sairaj18 commented on src/extended_metrics.go:
```diff
@@ -0,0 +1,77 @@
+package main
```
for now we are going ahead with `_` instead og `-`

sairaj18 commented on src/slave_metrics.go:
```diff
@@ -0,0 +1,74 @@
+package main
```
for now we are going ahead with `_` instead og `-`

sairaj18 commented:
> Thanks for taking care of this.
> 
> Left some comments. Some more points:
> 
> 1. Also, a lot of new functions are added without unit tests. I think it makes sense to add tests for the ones that have logical statements.
> 2. Do you think it improves the code if we have an optional argument in config with db version? We won't have to compute version on each execution. If not specified, we anyways do it.
> 3. IMO, the additional dependency is not required. But we can discuss.
> 4. Can `version` throughout the code be replaced with `dbVersion` or something similar. It could be confused with integration version.

1. Added unit tests for the new functions that have introduced.
2. We can check the optional argument configuration with kai
3. removed the additional dependency as we discussed
4. replaced `version` with `dbVersion` in most of the code wherever needed

rajrohanyadav reviewed (approved):
LGTM

Just one point as mentioned in my last comment. It also provides a way to not break the integration if we are unable to get version.
> Do you think it improves the code if we have an optional argument in config with db version? We won't have to compute version on each execution. If not specified, we anyways do it.

sairaj18 commented:
Tested with the latest changes - [here](https://new-relic.atlassian.net/browse/NR-133316?focusedCommentId=1095087)

rahulreddy15 reviewed (dismissed):
Suggestion: More details about the integration testing setup and how the Dockerfiles are setup and assumptions made while creating them could be added to the `README` in `tests/integration`. This might save time when we have to revisit this in the future.

LGTM 🚀

sairaj18 commented:
> Suggestion: More details about the integration testing setup and how the Dockerfiles are setup and assumptions made while creating them could be added to the `README` in `tests/integration`. This might save time when we have to revisit this in the future.
> 
> LGTM 🚀

Created Jira for this https://new-relic.atlassian.net/browse/NR-356321

rajrohanyadav reviewed (approved):
LGTM, thanks
